glean-core/android/src/test/java/mozilla/telemetry/glean/pings/CustomPingTest.kt

```diff
@@ -0,0 +1,109 @@
+import mozilla.telemetry.glean.private.NoReasonCodes
+import mozilla.telemetry.glean.private.PingType
+import mozilla.telemetry.glean.resetGlean
+import mozilla.telemetry.glean.testing.GleanTestRule
+import mozilla.telemetry.glean.triggerWorkManager
+import mozilla.telemetry.glean.utils.tryGetLong
+import org.junit.Assert.assertEquals
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.json.JSONObject
+import java.util.concurrent.TimeUnit
+
+/**
+ * Testing behavior of custom pings.
+ *
+ * We already rely on the Rust side to test custom pings,
+ * but this enables us to test the upload mechanism specifically.
+ *
+ * Even if this seemingly duplicates some of the testing, this should be kept around.
+ */
+@RunWith(AndroidJUnit4::class)
+class CustomPingTest {
+    @get:Rule
+    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())
+
+    @Test
+    fun `sends empty custom ping`() {
+        // a smoke test for custom pings
+
+        val server = getMockWebServer()
+        val context = getContextWithMockedInfo()
+
+        resetGlean(context, Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ), clearStores = true, uploadEnabled = true)
+
+        // Define a new custom ping inline.
+        val customPing = PingType<NoReasonCodes>(
+            name = "custom-ping",
+            includeClientId = true,
+            sendIfEmpty = true,
+            reasonCodes = emptyList()
+        )
+
+        customPing.submit()
+        triggerWorkManager(context)
+
+        val request = server.takeRequest(2L, TimeUnit.SECONDS)
+        val docType = request.path.split("/")[3]
+        assertEquals("custom-ping", docType)
+    }
+
+    @Test
+    fun `multiple pings in one go`() {
+        val server = getMockWebServer()
+        val context = getContextWithMockedInfo()
+
+        resetGlean(context, Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ), clearStores = true, uploadEnabled = true)
+
+        // Define a new custom ping inline.
+        val customPing = PingType<NoReasonCodes>(
+            name = "custom-ping",
+            includeClientId = true,
+            sendIfEmpty = true,
+            reasonCodes = emptyList()
+        )
+
+        // Trigger the ping twice.
+        customPing.submit()
+        customPing.submit()
+
+        // Trigger work manager once.
+        // This should launch one worker that handles all pending pings.
+        triggerWorkManager(context)
+
+        // Receive the first ping.
+        var request = server.takeRequest(2L, TimeUnit.SECONDS)
+        var docType = request.path.split("/")[3]
+        assertEquals("custom-ping", docType)
+
+        // Not much data in these pings,
+        // but order should be preserved, so we can check the sequence number.
```

Comment:
How is the order preserved?  Aren't they just filenames with a uuid on the filesystem?

Reply:
2 things:

1. While running they are added to the in-memory queue, thus preserving order
2. On startup we scan the directory and sort files by modified time.